[Malleability] Header

tim-barry opened the pull request:
Updates the `Header` type to make it non-malleable, by moving `ProposerSigData` field out of the Header to the new `ProposalHeader`/`BlockProposal`/`ClusterBlockProposal` types. `Header.ID()` will continue to return the same values as before this change.

`BlockProposal` is a new distinct type, used for proposing new blocks and receiving/verifying incoming proposed blocks from the network. It is an internal type corresponding to `messages.BlockProposal`/`messages.ClusterBlockProposal` (now `messages.UntrustedProposal` and `messages.UntrustedClusterProposal`), containing the header/block and associated `ProposerSigData`. The main changes associated with this are in the Compliance Core, consensus/collection Builder, and MessageHub; as well as adding some new conversion functions and unit test fixtures.

See #7191 for more specifics on usage and storage of Proposal signatures.

This PR preserves the "body" intermediate type used in the custom Fingerprint method to calculate the Header ID, which converts the Timestamp to unix time (necessary because `time.Time` is not RLP-encodable and therefore malleable) and flattens `LastViewTC` into an ID. The header timestamp and ID calculation is planned to later be changed in #7325 (#7209).
Various serialization methods and their tests are also preserved (TODO: discuss whether to remove JSON and/or Msgpack).

Closes #6656

durkmurder commented on model/flow/header.go:
```diff
@@ -39,29 +52,31 @@ type Header struct {
+// EncodableHeader is a type only used to calculate ID of a Header, by converting
+// Timestamp from time.Time to unix time (uint64) and LastViewTC to its identifier LastViewTCID.
+// This is necessary because time.Time is not RLP-encodable or decodable (due to having private fields).
+type EncodableHeader struct {
```
I think we can condense this into `Fingerprint` so basically something like this:

```
func (h Header) Fingerprint() []byte {
	return fingerprint.Fingerprint(struct {
		ChainID            ChainID
		ParentID           Identifier
		Height             uint64
		PayloadHash        Identifier
		Timestamp          uint64
		View               uint64
		ParentView         uint64
		ParentVoterIndices []byte
		ParentVoterSigData []byte
		ProposerID         Identifier
		LastViewTCID       Identifier
	}{
		ChainID:            h.ChainID,
		ParentID:           h.ParentID,
		Height:             h.Height,
		PayloadHash:        h.PayloadHash,
		Timestamp:          uint64(h.Timestamp.UnixNano()),
		View:               h.View,
		ParentView:         h.ParentView,
		ParentVoterIndices: h.ParentVoterIndices,
		ParentVoterSigData: h.ParentVoterSigData,
		ProposerID:         h.ProposerID,
		LastViewTCID:       h.LastViewTC.ID(),
	})
}

// ID returns a unique ID to singularly identify the header and its block
// within the flow system.
func (h Header) ID() Identifier {
	return MakeID(h)
}
```

tim-barry replied:
I originally made the change of creating an exported type (rather than anonymous inline struct) to avoid having two definitions of the same struct in different places, due to TestHeaderFingerprint unmarshaling the fingerprint:
https://github.com/onflow/flow-go/blob/485e20c9e6584a5acf2d8280768e67fbc46de568/model/flow/header_test.go#L36-L38
Not sure if that's a big concern; I do prefer not to have to export a type we don't really ever use, so I think your suggestion also works.

tim-barry replied:
Yeah, I think that the main reason for the test is now covered by the malleability checker test; so I don't think the RLP decoding test is necessary anymore.

tim-barry replied:
done in https://github.com/onflow/flow-go/pull/7100/commits/9e90359efb7225edb1e6db68f47ddda4f9ba423f

tim-barry commented:
I've determined at least some of the failing tests are due to a missing proposer signature in the `BlockResponse` to a `RangeRequest`, in the synchronization engine.
https://github.com/onflow/flow-go/blob/89dfb91165cf92181bfe2ece4bfc3afda11c7cce/engine/common/synchronization/engine.go#L317

The blocks that are requested and synced then go through all the same logic as new block proposals, so their proposer signature is checked and found to be missing.

Appears in logs as `[...] "error":"invalid proposal {ID} at view 1: invalid proposer signature: invalid vote at view 1 for block {ID}: could not split signature for block {ID}: invalid sig data length 0: signature's binary format is invalid","time":"2025-03-14T18:17:22Z","message":"received invalid block from other node (potential slashing evidence?)"}`

durkmurder reviewed (approved):
Very nice, great work. I think there are some open questions how do we deal with syncing of blocks and recovery, most likely we will need to store the signature either way but I think we should leave this for another PR and have a separate discussion about this.

tim-barry commented on module/builder/consensus/builder.go:
```diff
@@ -619,12 +620,12 @@ func (b *Builder) createProposal(parentID flow.Identifier,
-	sign func(*flow.Header) error,
-) (*flow.Block, error) {
+	sign func(*flow.Proposal) error,
+) (*flow.Block, *flow.Proposal, error) {
```
I also would prefer that, but the `builder.BuildOn` interface (and callers above that) is shared between Collection (cluster block) and Consensus (flow block), which have different proposal types due to having different payload types. The 'header proposal' type here is effectively a common denominator that can be used for both collection nodes and consensus nodes.
We could potentially introduce a new interface type instead for sharing this code between the LN and SN nodes, and then only have concrete `BlockProposal` types.

jordanschalm replied:
I think we _could_ use `BlockProposal` here per Yurii's suggestion, since this is internal to the implementation. We would still return `Proposal` from `BuildOn`. 

I'm also wondering if changing the `sign` function type would make working with the new types easier. For example, changing it to `sign func(*flow.header) ([]byte, error)` and just returning the signature for the input unsigned header. Then we can more easily use `BlockProposal` type internally here.

durkmurder replied:
> I'm also wondering if changing the `sign` function type would make working with the new types easier. For example, changing it to `sign func(*flow.header) ([]byte, error)` and just returning the signature for the input unsigned header. Then we can more easily use `BlockProposal` type internally here.

I was also thinking about this. We usually don't do this pattern in our code-base where we pass an object as argument to mutate it, instead we use the return argument, so I was thinking maybe there is a specific reason for this, but I couldn't find it. I would prefer to change `sign func(*flow.header) ([]byte, error)` as well.

tim-barry replied:
I agree that `sign func(*flow.header) ([]byte, error)` makes more sense; done in https://github.com/onflow/flow-go/pull/7100/commits/e75965cec7fb9afd991b9270e05a03028e438c80 and using `BlockProposal` done in https://github.com/onflow/flow-go/pull/7100/commits/349653b935f6912f3a41e5c7c7c0c4e8539bb09c

tim-barry commented on cmd/consensus/main.go:
```diff
@@ -793,10 +793,15 @@ func main() {
+			// TODO(tim) proposer signature storage
+			pending := make([]*flow.Proposal, 0, len(pendingOrig))
+			for _, p := range pendingOrig {
+				pending = append(pending, &flow.Proposal{Header: p, ProposerSigData: nil})
```
Yes I believe it does. This is one of the main reasons that we need storage for proposer signatures.

jordanschalm commented on engine/common/rpc/convert/headers.go:
```diff
@@ -48,7 +48,6 @@ func BlockHeaderToMessage(
 		ParentVoterIds:     parentVoterIds,
 		ParentVoterSigData: h.ParentVoterSigData,
 		ProposerId:         h.ProposerID[:],
-		ProposerSigData:    h.ProposerSigData,
```
We have decided to store the proposer signature for all node roles, at least for this round of changes. Given that, we can simply include the proposer signature here again.

AlexHentschel commented on model/messages/consensus.go:
```diff
@@ -129,12 +129,21 @@ func UntrustedBlockFromInternal(flowBlock *flow.Block) UntrustedBlock {
 // BlockProposal is part of the consensus protocol and represents the leader
 // of a consensus round pushing a new proposal to the network.
 type BlockProposal struct {
```
I got confused about the naming of `UntrustedX` vs `messages.X` when reviewing [PR #7191](https://github.com/onflow/flow-go/pull/7191), e.g. the diff on the file [model/messages/convert_test.go](https://github.com/onflow/flow-go/compare/65e37018c994aec7ce92ecbb046bb2ae6af8f802..e549d733c95c4d247a1aece538ae9388c96aa527#diff-276a1083c06ed7cf06983e0b6e520c8acd879865309f54622f886c4c8b86599c):

<img width="697" alt="Screenshot 2025-04-14 at 8 55 40 PM" src="https://github.com/user-attachments/assets/6cdd3f36-3570-4bd2-a5df-757097926fa3" />

I first was confused why we are removing the `Untrusted` qualifier, as this seems to be relatively important context in my opinion. Then I realized that we are not actually removing the word `Untrusted`. We are just switching to a different already-existing data structure in the same package, that does not have the `Untrusted` qualifier yet. 

Hence, I am thinking that we should add it be **renaming `BlockProposal` to `UntrustedProposal`**

tim-barry replied:
Done in https://github.com/onflow/flow-go/pull/7100/commits/f83b47e16295e9e468ce6395d920127c13c0c7c2

AlexHentschel commented on model/messages/collection.go:
```diff
@@ -75,12 +75,28 @@ func UntrustedClusterBlockFromInternal(clusterBlock *cluster.Block) UntrustedClu
 // consensus. The header contains information about consensus state and the
 // payload contains the proposed collection (may be empty).
 type ClusterBlockProposal struct {
```
Similarly to my previous comment, I would suggest that we **rename `ClusterBlockProposal` to `UntrustedClusterProposal`**

tim-barry replied:
Done in https://github.com/onflow/flow-go/pull/7100/commits/f83b47e16295e9e468ce6395d920127c13c0c7c2

tim-barry commented on engine/collection/compliance/core_test.go:
```diff
@@ -276,15 +272,16 @@ func (cs *CoreSuite) TestOnBlockProposal_FailsHotStuffValidation() {
-	proposal := messages.NewClusterBlockProposal(&block)
-	hotstuffProposal := model.SignedProposalFromFlow(block.Header)
+	proposal := unittest.ClusterProposalFromBlock(&block)
+	proposalMsg := messages.ClusterBlockProposalFrom(proposal) // TODO(tim) - unittest naming
```
resolved in https://github.com/onflow/flow-go/pull/7100/commits/f83b47e16295e9e468ce6395d920127c13c0c7c2 😁

tim-barry commented:
I believe Jordan's comment on the breaking API change (https://github.com/onflow/flow-go/pull/7100#discussion_r1999649490) is the only unaddressed comment on the Header changes (at least, that isn't already covered in #7311).

tim-barry commented:
Changes related to the API will be addressed in https://github.com/onflow/flow-go/issues/7331

jordanschalm reviewed (approved):
Reviewed the diff since https://github.com/onflow/flow-go/pull/7100 was merged, and the files where prior comments were addressed.

Great work getting this over the line. 🎸
